src/Microsoft.OData.Core/UriParser/SemanticAst/EachSegment.cs

```diff
@@ -0,0 +1,90 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="EachSegment.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.UriParser
+{
+    #region Namespaces
+    using Microsoft.OData.Edm;
+    #endregion Namespaces
+
+    /// <summary>
+    /// A segment representing $each in a path.
+    /// </summary>
+    public sealed class EachSegment : ODataPathSegment
+    {
+        /// <summary>
+        /// The <see cref="IEdmType"/> of the resource that this <see cref="EachSegment"/> represents.
+        /// </summary>
+        private readonly IEdmType edmType;
+
+        /// <summary>
+        /// Build a segment representing $each.
+        /// </summary>
+        /// <param name="navigationSource">The entity collection that this set-based operation applies to.</param>
+        /// <param name="targetEdmType">Target type for the entity being referenced.</param>
+        /// <exception cref="System.ArgumentNullException">Throws if any input parameter is null.</exception>
+        /// <remarks>$filter cannot be applied on singletons.</remarks>
+        public EachSegment(IEdmNavigationSource navigationSource, IEdmType targetEdmType)
+        {
+            ExceptionUtils.CheckArgumentNotNull(navigationSource, "navigationSource");
+
+            this.Identifier = UriQueryConstants.EachSegment;
+            this.SingleResult = false;
+            this.TargetEdmNavigationSource = navigationSource;
+            this.TargetKind = RequestTargetKind.Resource;
+            this.TargetEdmType = targetEdmType ?? navigationSource.EntityType();

```

Comment:
Note that, according to OData Protocol, $each is valid following any collection (not just an entity collection).  If we change this to get this.TargetKind based on targetEdmType, and this.TargetEdmType to be the passed in targetEdmType, is there anything else that assumes the collection being iterated is a collection of entities?

Reply:
Thanks for point this out. There aren't any other code paths that I wrote which would assume that the collection is a collection of entities. I've updated the line to:

`this.TargetKind = targetEdmType.GetTargetKindFromType();`

Reply:
Cool. Do we have any tests that validate the use against a collection(complex) or collection(primitive)?

Reply:
Adding two tests to explicitly test this:
```
        [Fact]
        public void SetBasedOperations_EachSegmentOnPrimitiveTypeCollection_ReturnsSuccess()
        {
            ParseUriAndVerify(
                new Uri("http://gobbledygook/People/1/RelatedIDs/$each"),
                (oDataPath, filterClause, aliasNodes) =>
                {
                    oDataPath.Count.Should().Be(4);

                    EachSegment eachSegment = oDataPath.Last() as EachSegment;
                    eachSegment.Should().NotBeNull();
                    eachSegment.TargetKind.Should().Be(RequestTargetKind.Primitive);
                    eachSegment.TargetEdmNavigationSource.Should().Be(HardCodedTestModel.GetPeopleSet());
                });
        }

        [Fact]
        public void SetBasedOperations_EachSegmentOnComplexTypeCollection_ReturnsSuccess()
        {
            ParseUriAndVerify(
                new Uri("http://gobbledygook/People/1/PreviousAddresses/$each"),
                (oDataPath, filterClause, aliasNodes) =>
                {
                    oDataPath.Count.Should().Be(4);

                    EachSegment eachSegment = oDataPath.Last() as EachSegment;
                    eachSegment.Should().NotBeNull();
                    eachSegment.TargetKind.Should().Be(RequestTargetKind.Resource);
                    eachSegment.TargetEdmNavigationSource.Should().Be(HardCodedTestModel.GetPeopleSet());
                });
        }
```